### ISSUE TITLE
Question: Async Routes with same key. Is that a bad pattern or am I doing something wrong?

### ISSUE THREAD

**Issue by terencechow** [davezuko/react-redux-starter-kit#1215]
I am using the fractal project structure which I like a lot.

However, I've come across a bit of a issue. I have a Show and an Edit route for a user. Both routes inject the same key i.e. `user` into the store.

When I go to the edit page and submit the form, it dispatches an update to the store, i.e. update `store.user.name` for example. After the successful update, I use `react-router` to go back to the show route. The whole function is shown below.

Unfortunately, when it pushes back to the Show page the store doesn't have the updated details for some reason. If I change the key that `injectReducer` uses so that it is a different key on each page, then the data does update. However I'd rather not do that because I like the modular nature of each route doing 1 thing and if I change the key, then I can't access the same `user` state in my action handlers.

Any reason why this is happening? Any suggestions? 

```
export function updateUser (data) {
  return (dispatch, getState) => {
    return fetchFEServer('PUT', 'updateUser', data)
    .then(r => r.json())
    .then((json) => {
      dispatch(updateUserSuccess(json.data.user))
    })
   .then(() =>  browserHistory.push(`/users/${json.data.user.id}`))
    .catch((err) => {
      console.log('update user error', err)
    })
  }
}
```

**Comment by terencechow** [comment]
Example [repo](https://github.com/terencechow/react-redux-starter-kit) showing my problem. 

Hitting Double Async will double the count and then go to another page with the same 'key' where you can Half Async which will bring you back to the original counter page. 

Note that I've commented out
`if (Object.hasOwnProperty.call(store.asyncReducers, key)) return` 
because I was using an older version of this starter kit that that did not have this.

What is confusing is that I see the sequence of events as

1) dispatch occurs -> (expect state to change)
2) browserHistory push occurs -> (expect page to change with new state)

What happens is:

dispatch occurs -> state does not change because store has already updated before the dispatch. 

It looks like the reducer has updated to the reducer on the following page BEFORE the dispatch occurs, hence nothing changes since there is no handler for that constant on the following page.

I also noticed that when going to the different pages, the state for that store is preserved, despite having the same key. It's almost like somewhere something is being cached.

**Comment by terencechow** [comment]
After much investigation I have discovered that in development ReduxDevTools will recompute the entire state history on a new page. Inserting a new reducer where `initialState` is different on the new page results in 2 different results. There are not 2 stores or any caches, it is just redux dev tools recomputing the entire state history with different `initialStates`.

Hence in my scenario, dispatch was updating the state on my first page. Then browserHistory makes a push to go to the second page. The second page recomputes the entire state history. However the second page has a reducer that is missing the action handler for the first page. Hence when  recomputing the entire state history, the state doesn't change from that last action.

Since all my reducers have the same signature, I recreated the async reducer with the following and instead of exporting reducers I export the action handlers. In production we only need to use the localHandlers and in dev, we create a special handlers object that contains all our handlers so that the new page can recompute with the correct handlers.

```
let handlers = {}
const createReducer = (key, localHandlers, initialState) => {
  handlers[key] = handlers[key] && __DEV__
    ? Object.assign({}, handlers[key], localHandlers)
    : localHandlers
  return (state = initialState, action) => {
    const handleFn = handlers[key][action.type]
    return handleFn ? handleFn(state, action) : state
  }
}

export const injectReducer = (store, { key, localHandlers }) => {
  const initialState = store.getState()[key]
  store.asyncReducers[key] = createReducer(key, localHandlers, initialState)
  store.replaceReducer(makeRootReducer(store.asyncReducers))
}
```
